types.ts

```diff
@@ -32,54 +32,88 @@ export enum DataTypeValue {
   stringType = 'stringType',
 }
 
-export interface DataType {
+export const DataTypeTemplateBase: z.ZodType<{
```

Comment:
what the zod is going on here??

Zod is being used to validate types/schemas (as all of TS's work is done at compile time only); as the scope of this expands our TS types need to get converted to Zod schemas. Zod does the same enforcement, with the bonus of runtime validation; additionally, we can still get the TS types (as we see exported here) using `z.infer`. The only time it gets a little ugly is recursion (see `DataTypeTemplateBase` and derivatives) and some times intellisense doesn't like to put our nice aliases (so we see ugly schemas instead of just `EntityTypeField`), but for now that's just how it has to be.